### ISSUE TITLE
Bonjour is not installed automatically for versions 4.0.0 and 4.1.0.

### ISSUE THREAD

**Issue by jason-mulree** [fttx/barcode-to-pc-server#371]
I have noticed that Bonjour is not installed automatically for versions 4.0.0 and 4.1.0 (64-bit, not sure if 32-bit versions are affected) and when these versions are launched, related error messages are displayed. This is despite running the installer with administrative privileges as advised by the error messages.

As a workaround, I installed version 3.18.1 which does install Bonjour, then removed this version (Bonjour remains installed when the old version is uninstalled) and installed version 4.1.0.

If you are quick enough, you can grab the 'Bonjour64.msi' file from your temporary folder during the installation of 3.18.1 so you can install Bonjour prior to installing the latest version for a 'cleaner' installation.

To replicate the issue:
1. Uninstall all versions of Barcode to PC server.
2. Uninstall Bonjour (this may cause problems with other applications that use Bonjour).
3. Download and install Barcode to PC server versions 4.0.0 or 4.1.0 (64-bit).
4. Notice that Bonjour is not installed automatically.
5. Notice the related error messages when launching Barcode to PC server.

Note: My operating is a fully updated Windows 10 (x64).

**Comment by jason-mulree** [comment]
My server is working fine without Bonjour so perhaps the installation of Bonjour should be optional?

Also, the Bonjour warnings are displayed when the application is configured to launch automatically on start, perhaps these warnings can be suppressed for the automatic start, especially if configured to launch minimised.

**Comment by fttx** [comment]
Thank you for the details. I've identified the issue.

Regarding the error, in my opinion, it's better to always show the error. One of the main goals of the app is to automatically connect to the server. So it's better to show an error to alert the user that something isn't working properly. The Bonjour package is often shared with other apps, so it isn't a big deal to have it installed on the system.

To reduce the user actions even more I'll also try to make it automatically configure Windows Firewall: #362

**Comment by fttx** [comment]
Ok, this is the installer with the new changes: https://drive.google.com/uc?export=download&id=1rPalC5uh47tL-tNIgCUYMNhYbWloCWLC

Could you please test it? I've also added a command to edit the Windows Firewall settings automatically. Initially, a black command-line window was opening for a few ms just after the installation, but I think I managed to hide it completely now.

Instructions:

Download the .zip using the link above
Extract everything in the same folder (important)
Open the .exe

**Comment by jason-mulree** [comment]
I am aware you have closed this issue and I apologise if my reply opens it again!

The new installer works well and Bonjour installs automatically as with 3.x versions. The application launches and functions as expected. When I installed other versions (not this test version), I am greeted with a Windows Firewall dialog to allow the application through the firewall - this adds the necessary rules without any problem, but probably good practice not to rely on the Windows Firewall to prompt, so I like that you have added a command to add the necessary rules to the Windows Firewall.

BTW, I think I caught a glimpse of Command Prompt window, but it was fleeting and I am not totally convinced that I saw it - in any case, I expect windows popping up briefly during an installation and consider it normal even if a little less than perfect, just because I am a perfectionist!
